arch/x86/kvm/kvmi.c

```diff
@@ -622,14 +622,277 @@ void kvmi_arch_breakpoint_event(struct kvm_vcpu *vcpu, u64 gva, u8 insn_len)
+{
+	size_t i, j;
+	unsigned short ***gfn_track;
+	unsigned long ***kvmi_track;
+
+	for (i = new; i < old; i++) {
+		for (j = 0; j < KVM_PAGE_TRACK_MAX; j++) {
+			kvfree(slot->arch.gfn_track[i][j]);
+			kvfree(slot->arch.kvmi_track[i][j]);
+		}
+
+		kvfree(slot->arch.gfn_track[i]);
+		kvfree(slot->arch.kvmi_track[i]);
+	}
+
+	gfn_track = krealloc(slot->arch.gfn_track,
+			     new * sizeof(short**), GFP_KERNEL_ACCOUNT);
+	if (!gfn_track)
+		return;
+	slot->arch.gfn_track = gfn_track;
+
+	kvmi_track = krealloc(slot->arch.kvmi_track,
+			      new * sizeof(long**), GFP_KERNEL_ACCOUNT);
+	if (!kvmi_track)
+		return;
+	slot->arch.kvmi_track = kvmi_track;
+}
+
+static bool grow_ept_tracks(struct kvm *kvm,
+			    u16 new)
+{
+	struct kvm_memory_slot *slot;
+	u16 old = kvm->arch.mmu_root_hpa_altviews_count;
+	size_t i;
+
+	for (i = 0; i < KVM_ADDRESS_SPACE_NUM; i++)
+		kvm_for_each_memslot(slot, __kvm_memslots(kvm, i))
+			if (!grow_ept_tracks_for_memslot(slot, old, new)) {
+				kvm->arch.mmu_root_hpa_altviews_count = new;
+				for (; i >= 0; i--)
+					shrink_ept_tracks_for_memslot(slot, new, old);
+				kvm->arch.mmu_root_hpa_altviews_count = old;
+				return false;
+			}
+
+	return true;
+}
+
+static void shrink_ept_tracks(struct kvm *kvm, u16 new)
+{
+	struct kvm_memory_slot *slot;
+	size_t i;
+
+	for (i = 0; i < KVM_ADDRESS_SPACE_NUM; i++)
+		kvm_for_each_memslot(slot, __kvm_memslots(kvm, i))
+			shrink_ept_tracks_for_memslot(slot, kvm->arch.mmu_root_hpa_altviews_count, new);
+}
+
+int kvmi_arch_cmd_create_ept_view(struct kvm *kvm)
+{
+	struct kvm_vcpu *vcpu;
+	struct hlist_head **page_hash;
+	struct kvm_mmu *mmu;
+	struct radix_tree_root *access_tree;
+	hpa_t *views;
+	u16 view = 0, views_count;
+	size_t i, j;
+
+	while (view < PTRS_PER_PGD - 1
+	       && kvm->arch.mmu_root_hpa_altviews_occupied[view])
+		++view;
+
+	if (++view >= PTRS_PER_PGD)
+		return -KVM_EINVAL;
+
+	views_count = view + 1;
+
+	if (!grow_ept_tracks(kvm, views_count))
+		return -KVM_ENOMEM;
+
+	kvm_for_each_vcpu(i, vcpu, kvm) {
+		mmu = vcpu->arch.mmu;
+		if (!mmu)
+			return -KVM_EINVAL;
+
+		if (views_count > kvm->arch.mmu_root_hpa_altviews_count) {
+			page_hash = krealloc(mmu->page_hash,
+					     views_count * sizeof(struct hlist_head *),
+					     GFP_KERNEL_ACCOUNT);
+			if (!page_hash)
+				return -KVM_ENOMEM;
+			for (j = kvm->arch.mmu_root_hpa_altviews_count; j < views_count; j++)
+				page_hash[j] = kzalloc(KVM_NUM_MMU_PAGES *
+						       sizeof(struct hlist_head),
+						       GFP_KERNEL_ACCOUNT);
+			mmu->page_hash = page_hash;
+
+			views = krealloc(mmu->root_hpa_altviews,
+					 views_count * sizeof(hpa_t), GFP_NOIO);
+			if (!views)
+				return -KVM_ENOMEM;
+			mmu->root_hpa_altviews = views;
+			mmu->root_hpa_altviews[view] = INVALID_PAGE;
+		}
+	}
+
+	if (views_count > kvm->arch.mmu_root_hpa_altviews_count) {
+		access_tree = krealloc(kvm->kvmi->access_tree,
+					views_count * sizeof(struct radix_tree_root),
+					GFP_KERNEL_ACCOUNT);
```

Comment:
I think you might need to use a write lock here: `write_lock(&kvm->kvmi->access_tree_lock)`. I haven't thoroughly checked this.

Reply:
I don't think this is necessary. `access_tree_lock` guards entries in the radix tree, but here we are just reallocating the roots. I don't think there should be any data races here since when we are growing/shrinking this array, there won't be any accesses to the tree before/after. Feel free to let me know if this thinking is misguided.